File: src/main/kotlin/Day1.kt
val words = arrayOf(
    "zero", "one", "two", "three", "four",
    "five", "six", "seven", "eight", "nine"
)

fun main(args: Array<String>) {
    println(readLines().sumOf { line ->
        var wordDigitsIndexes = words.flatMap { wordDigitsIndexes(line, it) }
        var digitIndexes = digitIndexes(line)
        val result = (wordDigitsIndexes + digitIndexes)
            .sortedBy { it.first }
        (result.first.second + result.last.second).toInt()
    })
}

fun wordDigitsIndexes(string: String, word: String): List<Pair<Int, String>> {
    var startIndex = 0
    val result = mutableListOf<Pair<Int, String>>()
    while (startIndex < string.length) {
        val indexOf = string.indexOf(word, startIndex)
        if (indexOf == -1) {
            break
        }
        result.add(Pair(indexOf, words.indexOf(word).toString()))
        startIndex = indexOf + 1
    }
    return result
}

fun digitIndexes(string: String): List<Pair<Int, String>> {
    val result = mutableListOf<Pair<Int, String>>()
    for ((index, char) in string.withIndex()) {
        if (char.isDigit()) {
            result.add(Pair(index, char.toString()))
        }
    }
    return result
}

File: day_1.py
import file_fetcher

words = [
    "zero", "one", "two", "three", "four",
    "five", "six", "seven", "eight", "nine"
]


def main():
    result = []
    lines = file_fetcher.fetch_all_lines()
    for line in lines:
        word_digit_indexes = sum([word_digits_indexes(line, word) for word in words], [])
        digit_indexes = digit_indexes_in_string(line)
        all_indexes = sorted(word_digit_indexes + digit_indexes, key=lambda x: x[0])
        result.append(int(all_indexes[0][1] + all_indexes[-1][1]))
    print(sum(result))


def word_digits_indexes(string, word):
    start_index = 0
    result = []
    while start_index < len(string):
        index_of = string.find(word, start_index)
        if index_of == -1:
            break
        result.append((index_of, str(words.index(word))))
        start_index = index_of + 1
    return result


def digit_indexes_in_string(string):
    return [(index, char) for index, char in enumerate(string) if char.isdigit()]


if __name__ == "__main__":
    main()